File: app/Sources/GunkApp/Analyze/SymbolExtractor.swift
import Foundation
import SwiftTreeSitter
import TreeSitterGo
import TreeSitterJavaScript
import TreeSitterPython
import TreeSitterSwift
import TreeSitterTypeScript

protocol SymbolExtractor {
  func extract(file: SymbolFile) throws -> FileSymbols
}

enum SymbolExtractorError: Error, Equatable {
  case parseFailed(String)
}

final class TreeSitterSymbolExtractor: SymbolExtractor {
  func extract(file: SymbolFile) throws -> FileSymbols {
    let language = LanguageKind(path: file.path)

    guard language != .unknown else {
      return fallbackExtract(file: file, language: language)
    }

    let root = try parseRoot(file: file, language: language)
    var symbols: [Symbol] = []
    var imports: [ImportRef] = []
    var exports: [ExportRef] = []

    walk(root) { node in
      let type = node.nodeType ?? ""
      let text = nodeText(node, in: file.contents)
      let line = lineNumber(for: node, in: file.contents)

      switch language {
      case .javaScript, .typeScript:
        collectJavaScriptLike(
          node: node,
          type: type,
          text: text,
          line: line,
          symbols: &symbols,
          imports: &imports,
          exports: &exports
        )
      case .python:
        collectPython(
          node: node,
          type: type,
          text: text,
          line: line,
          symbols: &symbols,
          imports: &imports
        )
      case .swift:
        collectSwift(
          node: node,
          type: type,
          text: text,
          line: line,
          symbols: &symbols,
          imports: &imports,
          exports: &exports
        )
      case .go:
        collectGo(
          node: node,
          type: type,
          text: text,
          line: line,
          symbols: &symbols,
          imports: &imports,
          exports: &exports
        )
      case .unknown:
        break
      }
    }

    return FileSymbols(
      path: file.path,
      language: language,
      symbols: symbols.uniqued(),
      imports: imports.uniqued(),
      exports: exports.uniqued()
    )
  }

  private func parseRoot(file: SymbolFile, language: LanguageKind) throws -> Node {
    let parser = Parser()

    switch language {
    case .go:
      try parser.setLanguage(Language(language: tree_sitter_go()))
    case .javaScript:
      try parser.setLanguage(Language(language: tree_sitter_javascript()))
    case .python:
      try parser.setLanguage(Language(language: tree_sitter_python()))
    case .swift:
      try parser.setLanguage(Language(language: tree_sitter_swift()))
    case .typeScript:
      try parser.setLanguage(Language(language: tree_sitter_typescript()))
    case .unknown:
      throw SymbolExtractorError.parseFailed(file.path)
    }

    guard let tree = parser.parse(file.contents),
          let root = tree.rootNode else {
      throw SymbolExtractorError.parseFailed(file.path)
    }

    return root
  }

  private func collectJavaScriptLike(
    node: Node,
    type: String,
    text: String,
    line: Int,
    symbols: inout [Symbol],
    imports: inout [ImportRef],
    exports: inout [ExportRef]
  ) {
    switch type {
    case "import_statement":
      for specifier in moduleSpecifiers(in: text) {
        imports.append(ImportRef(moduleSpecifier: specifier, resolvedTarget: relativeTarget(for: specifier), line: line))
      }
    case "call_expression":
      if let specifier = requireSpecifier(in: text) {
        imports.append(ImportRef(moduleSpecifier: specifier, resolvedTarget: relativeTarget(for: specifier), line: line))
      }
    case "function_declaration":
      appendSymbol(from: node, in: text, kind: .function, line: line, symbols: &symbols)
    case "class_declaration":
      appendSymbol(from: node, in: text, kind: .class, line: line, symbols: &symbols)
    case "method_definition":
      appendSymbol(from: node, in: text, kind: .method, line: line, symbols: &symbols)
    case "interface_declaration":
      appendSymbol(from: node, in: text, kind: .interface, line: line, symbols: &symbols)
    case "type_alias_declaration":
      appendSymbol(from: node, in: text, kind: .type, line: line, symbols: &symbols)
    case "lexical_declaration", "variable_declaration":
      for name in variableNames(in: text) {
        symbols.append(Symbol(name: name, kind: .variable, line: line))
      }
    case "export_statement":
      exports.append(contentsOf: javaScriptExports(in: text, line: line))
      for specifier in moduleSpecifiers(in: text) {
        imports.append(ImportRef(moduleSpecifier: specifier, resolvedTarget: relativeTarget(for: specifier), line: line))
      }
    default:
      break
    }
  }

  private func collectPython(
    node: Node,
    type: String,
    text: String,
    line: Int,
    symbols: inout [Symbol],
    imports: inout [ImportRef]
  ) {
    switch type {
    case "import_statement", "import_from_statement":
      for specifier in pythonImportSpecifiers(in: text) {
        imports.append(ImportRef(moduleSpecifier: specifier, resolvedTarget: nil, line: line))
      }
    case "function_definition":
      appendSymbol(from: node, in: text, kind: .function, line: line, symbols: &symbols)
    case "class_definition":
      appendSymbol(from: node, in: text, kind: .class, line: line, symbols: &symbols)
    default:
      break
    }
  }

  private func collectSwift(
    node: Node,
    type: String,
    text: String,
    line: Int,
    symbols: inout [Symbol],
    imports: inout [ImportRef],
    exports: inout [ExportRef]
  ) {
    switch type {
    case "import_declaration":
      if let specifier = firstMatch(in: text, pattern: #"import\s+(?:\w+\s+)?([A-Za-z_][A-Za-z0-9_]*)"#) {
        imports.append(ImportRef(moduleSpecifier: specifier, resolvedTarget: nil, line: line))
      }
    case "function_declaration":
      appendSwiftDeclaration(in: text, line: line, symbols: &symbols, exports: &exports)
    case "class_declaration":
      appendSwiftDeclaration(in: text, line: line, symbols: &symbols, exports: &exports)
    case "struct_declaration":
      appendSwiftDeclaration(in: text, line: line, symbols: &symbols, exports: &exports)
    case "enum_declaration":
      appendSwiftDeclaration(in: text, line: line, symbols: &symbols, exports: &exports)
    case "protocol_declaration":
      appendSwiftDeclaration(in: text, line: line, symbols: &symbols, exports: &exports)
    default:
      break
    }
  }

  private func collectGo(
    node: Node,
    type: String,
    text: String,
    line: Int,
    symbols: inout [Symbol],
    imports: inout [ImportRef],
    exports: inout [ExportRef]
  ) {
    switch type {
    case "import_spec":
      for specifier in moduleSpecifiers(in: text) {
        imports.append(ImportRef(moduleSpecifier: specifier, resolvedTarget: nil, line: line))
      }
    case "function_declaration":
      appendSymbol(from: node, in: text, kind: .function, line: line, symbols: &symbols)
    case "method_declaration":
      appendSymbol(from: node, in: text, kind: .method, line: line, symbols: &symbols)
    case "type_declaration":
      for name in goTypeNames(in: text) {
        let symbol = Symbol(name: name, kind: .type, line: line)
        symbols.append(symbol)
        if name.first?.isUppercase == true {
          exports.append(ExportRef(name: name, kind: .type, line: line))
        }
      }
    default:
      break
    }

    if let symbol = symbols.last,
       symbol.line == line,
       symbol.name.first?.isUppercase == true {
      exports.append(ExportRef(name: symbol.name, kind: symbol.kind, line: line))
    }
  }

  private func appendSymbol(from node: Node, in text: String, kind: Symbol.Kind, line: Int, symbols: inout [Symbol]) {
    guard let nameNode = node.child(byFieldName: "name") else {
      return
    }

    let nodeLocation = node.range.location
    let relativeRange = NSRange(location: nameNode.range.location - nodeLocation, length: nameNode.range.length)
    let nsString = text as NSString
    guard relativeRange.location >= 0,
          relativeRange.location + relativeRange.length <= nsString.length else {
      return
    }

    let name = nsString.substring(with: relativeRange)
    guard !name.isEmpty else {
      return
    }

    symbols.append(Symbol(name: name, kind: kind, line: line))
  }

  private func appendSwiftDeclaration(in text: String, line: Int, symbols: inout [Symbol], exports: inout [ExportRef]) {
    guard let symbol = swiftPrimaryDeclaration(in: text, line: line) else {
      return
    }

    symbols.append(symbol)

    if swiftPrimaryDeclarationIsExported(in: text) {
      exports.append(ExportRef(name: symbol.name, kind: symbol.kind, line: line))
    }
  }

  private func fallbackExtract(file: SymbolFile, language: LanguageKind) -> FileSymbols {
    let imports = fallbackImports(in: file.contents)
    let symbols = fallbackSymbols(in: file.contents)

    return FileSymbols(
      path: file.path,
      language: language,
      symbols: symbols,
      imports: imports,
      exports: []
    )
  }
}

private func walk(_ node: Node, visit: (Node) -> Void) {
  visit(node)

  for index in 0..<node.childCount {
    guard let child = node.child(at: index) else {
      continue
    }

    walk(child, visit: visit)
  }
}

private func nodeText(_ node: Node, in contents: String) -> String {
  guard !contents.isEmpty else {
    return node.sExpressionString ?? ""
  }

  let nsString = contents as NSString
  let range = node.range
  guard range.location >= 0,
        range.location + range.length <= nsString.length else {
    return ""
  }

  return nsString.substring(with: range)
}

private func lineNumber(for node: Node, in contents: String) -> Int {
  let nsString = contents as NSString
  let location = max(0, min(node.range.location, nsString.length))
  let prefix = nsString.substring(with: NSRange(location: 0, length: location))
  return prefix.reduce(1) { count, character in
    character == "\n" ? count + 1 : count
  }
}

private func moduleSpecifiers(in text: String) -> [String] {
  matches(in: text, pattern: #""([^"]+)"|'([^']+)'"#)
    .compactMap { $0.first(where: { !$0.isEmpty }) }
}

private func requireSpecifier(in text: String) -> String? {
  firstMatch(in: text, pattern: #"require\s*\(\s*["']([^"']+)["']\s*\)"#)
}

private func pythonImportSpecifiers(in text: String) -> [String] {
  if let fromModule = firstMatch(in: text, pattern: #"^from\s+([A-Za-z_][A-Za-z0-9_\.]*)\s+import\s+"#) {
    return [fromModule]
  }

  guard let imported = firstMatch(in: text, pattern: #"^import\s+(.+)$"#) else {
    return []
  }

  return imported
    .split(separator: ",")
    .map { part in
      part.trimmingCharacters(in: .whitespacesAndNewlines)
        .split(separator: " ")
        .first
        .map(String.init) ?? ""
    }
    .filter { !$0.isEmpty }
}

private func variableNames(in text: String) -> [String] {
  matches(in: text, pattern: #"(?:const|let|var)\s+([A-Za-z_$][A-Za-z0-9_$]*)"#)
    .compactMap(\.first)
}

private func javaScriptExports(in text: String, line: Int) -> [ExportRef] {
  var exports: [ExportRef] = []

  let patterns: [(String, Symbol.Kind)] = [
    (#"export\s+(?:default\s+)?function\s+([A-Za-z_$][A-Za-z0-9_$]*)"#, .function),
    (#"export\s+(?:default\s+)?class\s+([A-Za-z_$][A-Za-z0-9_$]*)"#, .class),
    (#"export\s+interface\s+([A-Za-z_$][A-Za-z0-9_$]*)"#, .interface),
    (#"export\s+type\s+([A-Za-z_$][A-Za-z0-9_$]*)"#, .type),
    (#"export\s+(?:const|let|var)\s+([A-Za-z_$][A-Za-z0-9_$]*)"#, .variable)
  ]

  for (pattern, kind) in patterns {
    exports.append(contentsOf: matches(in: text, pattern: pattern).compactMap(\.first).map {
      ExportRef(name: $0, kind: kind, line: line)
    })
  }

  if let namedExports = firstMatch(in: text, pattern: #"export\s*\{([^}]+)\}"#) {
    exports.append(
      contentsOf: namedExports
        .split(separator: ",")
        .map { part in
          part.trimmingCharacters(in: .whitespacesAndNewlines)
            .components(separatedBy: " as ")
            .last?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        }
        .filter { !$0.isEmpty }
        .map { ExportRef(name: $0, kind: nil, line: line) }
    )
  }

  return exports
}

private func swiftPrimaryDeclaration(in text: String, line: Int) -> Symbol? {
  guard let regex = try? NSRegularExpression(
    pattern: #"\b(func|class|struct|enum|protocol)\s+([A-Za-z_][A-Za-z0-9_]*)"#,
    options: []
  ) else {
    return nil
  }

  let nsString = text as NSString
  let range = NSRange(location: 0, length: nsString.length)
  guard let match = regex.firstMatch(in: text, range: range),
        match.numberOfRanges == 3,
        match.range(at: 1).location != NSNotFound,
        match.range(at: 2).location != NSNotFound else {
    return nil
  }

  let keyword = nsString.substring(with: match.range(at: 1))
  let name = nsString.substring(with: match.range(at: 2))

  let kind: Symbol.Kind
  switch keyword {
  case "func":
    kind = .function
  case "class":
    kind = .class
  case "struct":
    kind = .struct
  case "enum":
    kind = .enum
  case "protocol":
    kind = .protocolDecl
  default:
    return nil
  }

  return Symbol(name: name, kind: kind, line: line)
}

private func swiftPrimaryDeclarationIsExported(in text: String) -> Bool {
  guard let regex = try? NSRegularExpression(
    pattern: #"^\s*(?:@[A-Za-z_][A-Za-z0-9_]*(?:\([^)]*\))?\s*)*(?:public|open)\s+\b(?:func|class|struct|enum|protocol)\b"#,
    options: []
  ) else {
    return false
  }

  let nsString = text as NSString
  return regex.firstMatch(in: text, range: NSRange(location: 0, length: nsString.length)) != nil
}

private func goTypeNames(in text: String) -> [String] {
  matches(in: text, pattern: #"type\s+([A-Za-z_][A-Za-z0-9_]*)"#)
    .compactMap(\.first)
}

private func fallbackImports(in text: String) -> [ImportRef] {
  var imports: [ImportRef] = []
  let lines = text.components(separatedBy: .newlines)

  for (index, line) in lines.enumerated() {
    let lineNumber = index + 1
    for specifier in moduleSpecifiers(in: line) where line.contains("import") || line.contains("require") {
      imports.append(ImportRef(moduleSpecifier: specifier, resolvedTarget: relativeTarget(for: specifier), line: lineNumber))
    }

    if let pythonModule = firstMatch(in: line, pattern: #"^(?:from|import)\s+([A-Za-z_][A-Za-z0-9_\.]*)"#) {
      imports.append(ImportRef(moduleSpecifier: pythonModule, resolvedTarget: nil, line: lineNumber))
    }

    if let swiftModule = firstMatch(in: line, pattern: #"^import\s+([A-Za-z_][A-Za-z0-9_]*)"#) {
      imports.append(ImportRef(moduleSpecifier: swiftModule, resolvedTarget: nil, line: lineNumber))
    }
  }

  return imports.uniqued()
}

private func fallbackSymbols(in text: String) -> [Symbol] {
  var symbols: [Symbol] = []
  let lines = text.components(separatedBy: .newlines)

  for (index, line) in lines.enumerated() {
    let lineNumber = index + 1

    if let name = firstMatch(in: line, pattern: #"\bfunction\s+([A-Za-z_$][A-Za-z0-9_$]*)"#) {
      symbols.append(Symbol(name: name, kind: .function, line: lineNumber))
    }

    if let name = firstMatch(in: line, pattern: #"\bclass\s+([A-Za-z_$][A-Za-z0-9_$]*)"#) {
      symbols.append(Symbol(name: name, kind: .class, line: lineNumber))
    }
  }

  return symbols.uniqued()
}

private func relativeTarget(for specifier: String) -> String? {
  specifier.hasPrefix(".") ? specifier : nil
}

private func firstMatch(in text: String, pattern: String) -> String? {
  matches(in: text, pattern: pattern).first?.first
}

private func matches(in text: String, pattern: String) -> [[String]] {
  guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else {
    return []
  }

  let nsString = text as NSString
  let range = NSRange(location: 0, length: nsString.length)

  return regex.matches(in: text, range: range).map { match in
    (1..<match.numberOfRanges).compactMap { index in
      let range = match.range(at: index)
      guard range.location != NSNotFound else {
        return nil
      }

      return nsString.substring(with: range)
    }
  }
}

private extension Array where Element: Hashable {
  func uniqued() -> [Element] {
    var seen = Set<Element>()
    return filter { seen.insert($0).inserted }
  }
}

File: engine/src/analyze/symbolExtractor.ts
import { Language, Parser, type Node } from "web-tree-sitter";

// Embed the tree-sitter runtime + grammars into the build with `type: "file"`.
// Under `bun run` these resolve to node_modules paths; in the compiled
// `gunk-engine` binary Bun embeds them and hands back a runtime-extractable
// path, so the single binary needs no node_modules on the target machine.
import treeSitterRuntimeWasm from "web-tree-sitter/tree-sitter.wasm" with { type: "file" };
import javascriptWasm from "tree-sitter-wasms/out/tree-sitter-javascript.wasm" with { type: "file" };
import typescriptWasm from "tree-sitter-wasms/out/tree-sitter-typescript.wasm" with { type: "file" };
import tsxWasm from "tree-sitter-wasms/out/tree-sitter-tsx.wasm" with { type: "file" };
import pythonWasm from "tree-sitter-wasms/out/tree-sitter-python.wasm" with { type: "file" };
import swiftWasm from "tree-sitter-wasms/out/tree-sitter-swift.wasm" with { type: "file" };
import goWasm from "tree-sitter-wasms/out/tree-sitter-go.wasm" with { type: "file" };

import {
  type ExportRef,
  type FileSymbols,
  type ImportRef,
  type LanguageKind,
  languageKindForPath,
  type Symbol,
  type SymbolKind,
} from "../models.js";
import { analysisFirstMatch, analysisMatches, analysisRawMatches } from "./analysisRegex.js";

export interface SymbolFile {
  path: string;
  contents: string;
}

export interface SymbolExtractor {
  extract(file: SymbolFile): FileSymbols;
}

type GrammarName = "javascript" | "typescript" | "tsx" | "python" | "swift" | "go";

const GRAMMAR_WASM: Record<GrammarName, string> = {
  javascript: javascriptWasm,
  typescript: typescriptWasm,
  tsx: tsxWasm,
  python: pythonWasm,
  swift: swiftWasm,
  go: goWasm,
};

const GRAMMARS: GrammarName[] = ["javascript", "typescript", "tsx", "python", "swift", "go"];

// --- structural dedupe (parity with Swift `Array.uniqued()` over Hashable) ---

function uniquedBy<T>(items: T[], key: (item: T) => string): T[] {
  const seen = new Set<string>();
  const out: T[] = [];
  for (const item of items) {
    const k = key(item);
    if (!seen.has(k)) {
      seen.add(k);
      out.push(item);
    }
  }
  return out;
}

function symbolKey(symbol: Symbol): string {
  return `${symbol.name}\u0000${symbol.kind}\u0000${symbol.line}`;
}

function importKey(ref: ImportRef): string {
  return `${ref.moduleSpecifier}\u0000${ref.resolvedTarget ?? "\u0001"}\u0000${ref.line}`;
}

function exportKey(ref: ExportRef): string {
  return `${ref.name}\u0000${ref.kind ?? "\u0001"}\u0000${ref.line}`;
}

function firstCharIsUppercase(name: string): boolean {
  if (name.length === 0) {
    return false;
  }
  const first = name[0];
  return first !== first.toLowerCase() && first === first.toUpperCase();
}

// --- regex helpers (faithful ports of the private Swift functions) ---

function moduleSpecifiers(text: string): string[] {
  return analysisMatches(text, String.raw`"([^"]+)"|'([^']+)'`)
    .map((groups) => groups.find((value) => value.length > 0))
    .filter((value): value is string => value !== undefined);
}

function requireSpecifier(text: string): string | undefined {
  return analysisFirstMatch(text, String.raw`require\s*\(\s*["']([^"']+)["']\s*\)`);
}

function pythonImportSpecifiers(text: string): string[] {
  const fromModule = analysisFirstMatch(
    text,
    String.raw`^from\s+([A-Za-z_][A-Za-z0-9_\.]*)\s+import\s+`,
  );
  if (fromModule !== undefined) {
    return [fromModule];
  }

  const imported = analysisFirstMatch(text, String.raw`^import\s+(.+)$`);
  if (imported === undefined) {
    return [];
  }

  return imported
    .split(",")
    .map((part) => {
      const trimmed = part.trim();
      const tokens = trimmed.split(" ").filter((token) => token.length > 0);
      return tokens.length > 0 ? tokens[0] : "";
    })
    .filter((value) => value.length > 0);
}

function variableNames(text: string): string[] {
  return analysisMatches(text, String.raw`(?:const|let|var)\s+([A-Za-z_$][A-Za-z0-9_$]*)`)
    .map((groups) => groups[0])
    .filter((value): value is string => value !== undefined);
}

function javaScriptExports(text: string, line: number): ExportRef[] {
  const exports: ExportRef[] = [];

  const patterns: [string, SymbolKind][] = [
    [String.raw`export\s+(?:default\s+)?function\s+([A-Za-z_$][A-Za-z0-9_$]*)`, "function"],
    [String.raw`export\s+(?:default\s+)?class\s+([A-Za-z_$][A-Za-z0-9_$]*)`, "class"],
    [String.raw`export\s+interface\s+([A-Za-z_$][A-Za-z0-9_$]*)`, "interface"],
    [String.raw`export\s+type\s+([A-Za-z_$][A-Za-z0-9_$]*)`, "type"],
    [String.raw`export\s+(?:const|let|var)\s+([A-Za-z_$][A-Za-z0-9_$]*)`, "variable"],
  ];

  for (const [pattern, kind] of patterns) {
    for (const groups of analysisMatches(text, pattern)) {
      const name = groups[0];
      if (name !== undefined) {
        exports.push({ name, kind, line });
      }
    }
  }

  const namedExports = analysisFirstMatch(text, String.raw`export\s*\{([^}]+)\}`);
  if (namedExports !== undefined) {
    for (const part of namedExports.split(",")) {
      const segments = part.trim().split(" as ");
      const name = (segments[segments.length - 1] ?? "").trim();
      if (name.length > 0) {
        exports.push({ name, kind: null, line });
      }
    }
  }

  return exports;
}

function swiftPrimaryDeclaration(text: string, line: number): Symbol | null {
  const matches = analysisMatches(
    text,
    String.raw`\b(func|class|struct|enum|protocol)\s+([A-Za-z_][A-Za-z0-9_]*)`,
  );
  const first = matches[0];
  if (!first || first.length < 2) {
    return null;
  }

  const keyword = first[0];
  const name = first[1];

  let kind: SymbolKind;
  switch (keyword) {
    case "func":
      kind = "function";
      break;
    case "class":
      kind = "class";
      break;
    case "struct":
      kind = "struct";
      break;
    case "enum":
      kind = "enum";
      break;
    case "protocol":
      kind = "protocolDecl";
      break;
    default:
      return null;
  }

  return { name, kind, line };
}

function swiftPrimaryDeclarationIsExported(text: string): boolean {
  return (
    analysisRawMatches(
      text,
      String.raw`^\s*(?:@[A-Za-z_][A-Za-z0-9_]*(?:\([^)]*\))?\s*)*(?:public|open)\s+\b(?:func|class|struct|enum|protocol)\b`,
    ).length > 0
  );
}

function goTypeNames(text: string): string[] {
  return analysisMatches(text, String.raw`type\s+([A-Za-z_][A-Za-z0-9_]*)`)
    .map((groups) => groups[0])
    .filter((value): value is string => value !== undefined);
}

function relativeTarget(specifier: string): string | null {
  return specifier.startsWith(".") ? specifier : null;
}

function splitLines(text: string): string[] {
  return text.split(/\r\n|\r|\n/);
}

function fallbackImports(text: string): ImportRef[] {
  const imports: ImportRef[] = [];
  const lines = splitLines(text);

  lines.forEach((line, index) => {
    const lineNumber = index + 1;
    if (line.includes("import") || line.includes("require")) {
      for (const specifier of moduleSpecifiers(line)) {
        imports.push({
          moduleSpecifier: specifier,
          resolvedTarget: relativeTarget(specifier),
          line: lineNumber,
        });
      }
    }

    const pythonModule = analysisFirstMatch(
      line,
      String.raw`^(?:from|import)\s+([A-Za-z_][A-Za-z0-9_\.]*)`,
    );
    if (pythonModule !== undefined) {
      imports.push({ moduleSpecifier: pythonModule, resolvedTarget: null, line: lineNumber });
    }

    const swiftModule = analysisFirstMatch(line, String.raw`^import\s+([A-Za-z_][A-Za-z0-9_]*)`);
    if (swiftModule !== undefined) {
      imports.push({ moduleSpecifier: swiftModule, resolvedTarget: null, line: lineNumber });
    }
  });

  return uniquedBy(imports, importKey);
}

function fallbackSymbols(text: string): Symbol[] {
  const symbols: Symbol[] = [];
  const lines = splitLines(text);

  lines.forEach((line, index) => {
    const lineNumber = index + 1;

    const fnName = analysisFirstMatch(line, String.raw`\bfunction\s+([A-Za-z_$][A-Za-z0-9_$]*)`);
    if (fnName !== undefined) {
      symbols.push({ name: fnName, kind: "function", line: lineNumber });
    }

    const className = analysisFirstMatch(line, String.raw`\bclass\s+([A-Za-z_$][A-Za-z0-9_$]*)`);
    if (className !== undefined) {
      symbols.push({ name: className, kind: "class", line: lineNumber });
    }
  });

  return uniquedBy(symbols, symbolKey);
}

// --- tree walking ---

function walk(node: Node, visit: (node: Node) => void): void {
  visit(node);
  for (let index = 0; index < node.childCount; index += 1) {
    const child = node.child(index);
    if (child) {
      walk(child, visit);
    }
  }
}

function appendSymbol(
  node: Node,
  kind: SymbolKind,
  line: number,
  symbols: Symbol[],
): void {
  const nameNode = node.childForFieldName("name");
  if (!nameNode) {
    return;
  }

  const name = nameNode.text;
  if (name.length === 0) {
    return;
  }

  symbols.push({ name, kind, line });
}

function appendSwiftDeclaration(
  text: string,
  line: number,
  symbols: Symbol[],
  exports: ExportRef[],
): void {
  const symbol = swiftPrimaryDeclaration(text, line);
  if (!symbol) {
    return;
  }

  symbols.push(symbol);

  if (swiftPrimaryDeclarationIsExported(text)) {
    exports.push({ name: symbol.name, kind: symbol.kind, line });
  }
}

function collectJavaScriptLike(
  node: Node,
  type: string,
  text: string,
  line: number,
  symbols: Symbol[],
  imports: ImportRef[],
  exports: ExportRef[],
): void {
  switch (type) {
    case "import_statement":
      for (const specifier of moduleSpecifiers(text)) {
        imports.push({
          moduleSpecifier: specifier,
          resolvedTarget: relativeTarget(specifier),
          line,
        });
      }
      break;
    case "call_expression": {
      const specifier = requireSpecifier(text);
      if (specifier !== undefined) {
        imports.push({
          moduleSpecifier: specifier,
          resolvedTarget: relativeTarget(specifier),
          line,
        });
      }
      break;
    }
    case "function_declaration":
      appendSymbol(node, "function", line, symbols);
      break;
    case "class_declaration":
      appendSymbol(node, "class", line, symbols);
      break;
    case "method_definition":
      appendSymbol(node, "method", line, symbols);
      break;
    case "interface_declaration":
      appendSymbol(node, "interface", line, symbols);
      break;
    case "type_alias_declaration":
      appendSymbol(node, "type", line, symbols);
      break;
    case "lexical_declaration":
    case "variable_declaration":
      for (const name of variableNames(text)) {
        symbols.push({ name, kind: "variable", line });
      }
      break;
    case "export_statement":
      exports.push(...javaScriptExports(text, line));
      for (const specifier of moduleSpecifiers(text)) {
        imports.push({
          moduleSpecifier: specifier,
          resolvedTarget: relativeTarget(specifier),
          line,
        });
      }
      break;
    default:
      break;
  }
}

function collectPython(
  node: Node,
  type: string,
  text: string,
  line: number,
  symbols: Symbol[],
  imports: ImportRef[],
): void {
  switch (type) {
    case "import_statement":
    case "import_from_statement":
      for (const specifier of pythonImportSpecifiers(text)) {
        imports.push({ moduleSpecifier: specifier, resolvedTarget: null, line });
      }
      break;
    case "function_definition":
      appendSymbol(node, "function", line, symbols);
      break;
    case "class_definition":
      appendSymbol(node, "class", line, symbols);
      break;
    default:
      break;
  }
}

function collectSwift(
  type: string,
  text: string,
  line: number,
  symbols: Symbol[],
  imports: ImportRef[],
  exports: ExportRef[],
): void {
  switch (type) {
    case "import_declaration": {
      const specifier = analysisFirstMatch(
        text,
        String.raw`import\s+(?:\w+\s+)?([A-Za-z_][A-Za-z0-9_]*)`,
      );
      if (specifier !== undefined) {
        imports.push({ moduleSpecifier: specifier, resolvedTarget: null, line });
      }
      break;
    }
    case "function_declaration":
    case "class_declaration":
    case "struct_declaration":
    case "enum_declaration":
    case "protocol_declaration":
      appendSwiftDeclaration(text, line, symbols, exports);
      break;
    default:
      break;
  }
}

function collectGo(
  node: Node,
  type: string,
  text: string,
  line: number,
  symbols: Symbol[],
  imports: ImportRef[],
  exports: ExportRef[],
): void {
  switch (type) {
    case "import_spec":
      for (const specifier of moduleSpecifiers(text)) {
        imports.push({ moduleSpecifier: specifier, resolvedTarget: null, line });
      }
      break;
    case "function_declaration":
      appendSymbol(node, "function", line, symbols);
      break;
    case "method_declaration":
      appendSymbol(node, "method", line, symbols);
      break;
    case "type_declaration":
      for (const name of goTypeNames(text)) {
        symbols.push({ name, kind: "type", line });
        if (firstCharIsUppercase(name)) {
          exports.push({ name, kind: "type", line });
        }
      }
      break;
    default:
      break;
  }

  const last = symbols[symbols.length - 1];
  if (last && last.line === line && firstCharIsUppercase(last.name)) {
    exports.push({ name: last.name, kind: last.kind, line });
  }
}

function fallbackExtract(file: SymbolFile, language: LanguageKind): FileSymbols {
  return {
    path: file.path,
    language,
    symbols: fallbackSymbols(file.contents),
    imports: fallbackImports(file.contents),
    exports: [],
  };
}

function grammarFor(language: LanguageKind, path: string): GrammarName | null {
  switch (language) {
    case "go":
      return "go";
    case "javaScript":
      return "javascript";
    case "python":
      return "python";
    case "swift":
      return "swift";
    case "typeScript":
      return path.toLowerCase().endsWith(".tsx") ? "tsx" : "typescript";
    case "unknown":
      return null;
  }
}

class TreeSitterSymbolExtractor implements SymbolExtractor {
  private readonly languages: Map<GrammarName, Language>;

  constructor(languages: Map<GrammarName, Language>) {
    this.languages = languages;
  }

  extract(file: SymbolFile): FileSymbols {
    const language = languageKindForPath(file.path);

    if (language === "unknown") {
      return fallbackExtract(file, language);
    }

    const grammarName = grammarFor(language, file.path);
    const grammar = grammarName ? this.languages.get(grammarName) : undefined;
    if (!grammar) {
      return fallbackExtract(file, language);
    }

    const parser = new Parser();
    parser.setLanguage(grammar);
    const tree = parser.parse(file.contents);
    const root = tree?.rootNode;
    if (!root) {
      return fallbackExtract(file, language);
    }

    const symbols: Symbol[] = [];
    const imports: ImportRef[] = [];
    const exports: ExportRef[] = [];

    walk(root, (node) => {
      const type = node.type;
      const text = node.text;
      const line = node.startPosition.row + 1;

      switch (language) {
        case "javaScript":
        case "typeScript":
          collectJavaScriptLike(node, type, text, line, symbols, imports, exports);
          break;
        case "python":
          collectPython(node, type, text, line, symbols, imports);
          break;
        case "swift":
          collectSwift(type, text, line, symbols, imports, exports);
          break;
        case "go":
          collectGo(node, type, text, line, symbols, imports, exports);
          break;
      }
    });

    return {
      path: file.path,
      language,
      symbols: uniquedBy(symbols, symbolKey),
      imports: uniquedBy(imports, importKey),
      exports: uniquedBy(exports, exportKey),
    };
  }
}

/** Synchronous regex-only extractor for unknown languages or grammar-load failures. */
export class FallbackSymbolExtractor implements SymbolExtractor {
  extract(file: SymbolFile): FileSymbols {
    return fallbackExtract(file, languageKindForPath(file.path));
  }
}

let parserInitialized: Promise<void> | null = null;

function ensureParserInitialized(): Promise<void> {
  if (!parserInitialized) {
    parserInitialized = Parser.init({ locateFile: () => treeSitterRuntimeWasm });
  }
  return parserInitialized;
}

export async function createTreeSitterSymbolExtractor(): Promise<SymbolExtractor> {
  await ensureParserInitialized();

  const languages = new Map<GrammarName, Language>();
  for (const name of GRAMMARS) {
    languages.set(name, await Language.load(GRAMMAR_WASM[name]));
  }

  return new TreeSitterSymbolExtractor(languages);
}
